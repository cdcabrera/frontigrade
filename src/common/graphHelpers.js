import moment from 'moment';
import apiTypes from '../constants/apiConstants';
import helpers from './helpers';

const calculateGraphTotals = ({ dailyUsage, totalInstancesOpenshift, totalInstancesRhel }) => {
  const openshiftMemoryTime = dailyUsage.reduce(
    (acc, val) => acc + Number.parseFloat(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_OPENSHIFT_MEMORY]),
    0
  );

  const openshiftRuntimeTime = dailyUsage.reduce(
    (acc, val) => acc + Number.parseFloat(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_OPENSHIFT_RUNTIME]),
    0
  );

  const openshiftVcpuTime = dailyUsage.reduce(
    (acc, val) => acc + Number.parseFloat(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_OPENSHIFT_VCPU]),
    0
  );

  const rhelMemoryTime = dailyUsage.reduce(
    (acc, val) => acc + Number.parseFloat(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_RHEL_MEMORY]),
    0
  );

  const rhelRuntimeTime = dailyUsage.reduce(
    (acc, val) => acc + Number.parseFloat(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_RHEL_RUNTIME]),
    0
  );

  const rhelVcpuTime = dailyUsage.reduce(
    (acc, val) => acc + Number.parseFloat(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_RHEL_VCPU]),
    0
  );

  return {
    instancesOpenshift: totalInstancesOpenshift,
    instancesRhel: totalInstancesRhel,
    openshiftMemoryTime: helpers.generateHoursFromSeconds(openshiftMemoryTime).hours,
    openshiftRuntimeTime: helpers.generateHoursFromSeconds(openshiftRuntimeTime).hours,
    openshiftVcpuTime: helpers.generateHoursFromSeconds(openshiftVcpuTime).hours,
    rhelMemoryTime: helpers.generateHoursFromSeconds(rhelMemoryTime).hours,
    rhelRuntimeTime: helpers.generateHoursFromSeconds(rhelRuntimeTime).hours,
    rhelVcpuTime: helpers.generateHoursFromSeconds(rhelVcpuTime).hours
  };
};

const convertGraphData = ({ dailyUsage }) => {
  const graphData = {
    date: ['x'],
    tooltipTitle: [],
    openshiftInstances: ['openshiftInstances'],
    openshiftMemoryTime: ['openshiftMemoryTime'],
    openshiftRuntimeTime: ['openshiftRuntimeTime'],
    openshiftVcpuTime: ['openshiftVcpuTime'],
    rhelInstances: ['rhelInstances'],
    rhelMemoryTime: ['rhelMemoryTime'],
    rhelRuntimeTime: ['rhelRuntimeTime'],
    rhelVcpuTime: ['rhelVcpuTime']
  };

  dailyUsage.forEach(val => {
    const formattedDate = moment
      .utc(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_DATE])
      .local()
      .format('YYYY-MM-DD');

    graphData.tooltipTitle.push(formattedDate);
    graphData.date.push(formattedDate);

    graphData.openshiftInstances.push(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_OPENSHIFT]);

    graphData.openshiftMemoryTime.push(
      helpers.generateHoursFromSeconds(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_OPENSHIFT_MEMORY]).hours
    );
    graphData.openshiftRuntimeTime.push(
      helpers.generateHoursFromSeconds(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_OPENSHIFT_RUNTIME]).hours
    );
    graphData.openshiftVcpuTime.push(
      helpers.generateHoursFromSeconds(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_OPENSHIFT_VCPU]).hours
    );

    graphData.rhelInstances.push(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_RHEL]);

    graphData.rhelMemoryTime.push(
      helpers.generateHoursFromSeconds(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_RHEL_MEMORY]).hours
    );
    graphData.rhelRuntimeTime.push(
      helpers.generateHoursFromSeconds(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_RHEL_RUNTIME]).hours
    );
    graphData.rhelVcpuTime.push(
      helpers.generateHoursFromSeconds(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_RHEL_VCPU]).hours
    );
  });

  const rhelInstances = {
    names: {
      rhelInstances: 'RHEL Instances'
    },
    columns: [graphData.date, graphData.rhelInstances],
    colors: { rhelInstances: helpers.pfPaletteColors.orange }
  };

  const openshiftInstances = {
    names: {
      openshiftInstances: 'RHOCP Instances'
    },
    columns: [graphData.date, graphData.openshiftInstances],
    colors: { openshiftInstances: helpers.pfPaletteColors.blue300 }
  };

  const openshiftMemoryTime = {
    names: {
      openshiftMemoryTime: 'RHOCP GB Memory Hours'
    },
    columns: [graphData.date, graphData.openshiftMemoryTime],
    colors: { openshiftMemoryTime: helpers.pfPaletteColors.blue300 }
  };

  const openshiftRuntimeTime = {
    names: {
      openshiftRuntimeTime: 'RHOCP Instance Hours'
    },
    columns: [graphData.date, graphData.openshiftRuntimeTime],
    colors: { openshiftRuntimeTime: helpers.pfPaletteColors.blue300 }
  };

  const openshiftVcpuTime = {
    names: {
      openshiftVcpuTime: 'RHOCP Core Hours'
    },
    columns: [graphData.date, graphData.openshiftVcpuTime],
    colors: { openshiftVcpuTime: helpers.pfPaletteColors.blue300 }
  };

  const rhelMemoryTime = {
    names: {
      rhelMemoryTime: 'RHEL GB Memory Hours'
    },
    columns: [graphData.date, graphData.rhelMemoryTime],
    colors: { rhelMemoryTime: helpers.pfPaletteColors.orange }
  };

  const rhelRuntimeTime = {
    names: {
      rhelRuntimeTime: 'RHEL Instance Hours'
    },
    columns: [graphData.date, graphData.rhelRuntimeTime],
    colors: { rhelRuntimeTime: helpers.pfPaletteColors.orange }
  };

  const rhelVcpuTime = {
    names: {
      rhelVcpuTime: 'RHEL Core Hours'
    },
    columns: [graphData.date, graphData.rhelVcpuTime],
    colors: { rhelVcpuTime: helpers.pfPaletteColors.orange }
  };

  const tooltips = {
    format: {
      title: d => graphData.tooltipTitle[d] || `Data ${d}`
    },
    tooltip: {
      grouped: false
    }
  };

  return {
    openshiftInstances,
    rhelInstances,
    openshiftMemoryTime,
    openshiftRuntimeTime,
    openshiftVcpuTime,
    rhelMemoryTime,
    rhelRuntimeTime,
    rhelVcpuTime,
    tooltips
  };
};

const graphHelpers = { calculateGraphTotals, convertGraphData };

export { graphHelpers as default, graphHelpers, calculateGraphTotals, convertGraphData };
