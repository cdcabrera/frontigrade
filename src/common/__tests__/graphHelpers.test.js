import { graphHelpers, calculateGraphTotals, convertGraphData } from '../graphHelpers';

describe('GraphHelpers', () => {
  const exampleGraphData = {
    dailyUsage: [
      {
        date: null,
        openshift_instances: 0,
        openshift_runtime_seconds: 0.0,
        rhel_instances: 1,
        rhel_runtime_seconds: 54000.0
      },
      {
        date: null,
        openshift_instances: 0,
        openshift_runtime_seconds: 0.0,
        rhel_instances: 1,
        rhel_runtime_seconds: 500.0
      },
      {
        date: null,
        openshift_instances: 0,
        openshift_runtime_seconds: 0.0,
        rhel_instances: 2,
        rhel_runtime_seconds: 9600.0
      },
      {
        date: null,
        openshift_instances: 1,
        openshift_runtime_seconds: 8000.0,
        rhel_instances: 2,
        rhel_runtime_seconds: 9500.0
      },
      {
        date: null,
        openshift_instances: 0,
        openshift_runtime_seconds: 0.0,
        rhel_instances: 2,
        rhel_runtime_seconds: 10000.0
      },
      {
        date: null,
        openshift_instances: 0,
        openshift_runtime_seconds: 0.0,
        rhel_instances: 2,
        rhel_runtime_seconds: 7600.0
      }
    ]
  };

  it('should export the same number of methods as imported', () => {
    expect(Object.keys(graphHelpers)).toHaveLength(2);
  });

  it('should have defined methods', () => {
    expect(graphHelpers.calculateGraphTotals).toBeDefined();

    const graphTotals = calculateGraphTotals({ ...exampleGraphData });
    expect(Object.keys(graphTotals)).toMatchSnapshot('calculateGraphTotals keys');

    expect(graphHelpers.convertGraphData).toBeDefined();

    const graphData = convertGraphData({ ...exampleGraphData });
    expect(Object.keys(graphData)).toMatchSnapshot('convertGraphData keys');
  });

  it('should consistently calculate totals', () => {
    expect(calculateGraphTotals({ ...exampleGraphData })).toMatchSnapshot('totals');
  });

  it('should consistently convert data', () => {
    expect(convertGraphData({ ...exampleGraphData })).toMatchSnapshot('converted');
  });

  it('should have a specific convertGraphData methods for tooltip titles', () => {
    const graphData = convertGraphData({ ...exampleGraphData });
    expect(graphData.tooltips.format.title('Lorem Ipsum')).toMatchSnapshot('tooltip title');
  });
});
