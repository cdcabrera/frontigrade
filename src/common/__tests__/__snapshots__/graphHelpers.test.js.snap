// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphHelpers should consistently calculate totals: totals 1`] = `
Object {
  "instancesOpenshift": undefined,
  "instancesRhel": undefined,
  "openshiftMemoryTime": null,
  "openshiftRuntimeTime": 3,
  "openshiftVcpuTime": null,
  "rhelMemoryTime": null,
  "rhelRuntimeTime": 26,
  "rhelVcpuTime": null,
}
`;

exports[`GraphHelpers should consistently convert data: converted 1`] = `
Object {
  "openshiftInstances": Object {
    "colors": Object {
      "openshiftInstances": "#39a5dc",
    },
    "columns": Array [
      Array [
        "x",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
      ],
      Array [
        "openshiftInstances",
        0,
        0,
        0,
        1,
        0,
        0,
      ],
    ],
    "names": Object {
      "openshiftInstances": "RHOCP Instances",
    },
  },
  "openshiftMemoryTime": Object {
    "colors": Object {
      "openshiftMemoryTime": "#39a5dc",
    },
    "columns": Array [
      Array [
        "x",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
      ],
      Array [
        "openshiftMemoryTime",
        null,
        null,
        null,
        null,
        null,
        null,
      ],
    ],
    "names": Object {
      "openshiftMemoryTime": "RHOCP GB Memory Hours",
    },
  },
  "openshiftRuntimeTime": Object {
    "colors": Object {
      "openshiftRuntimeTime": "#39a5dc",
    },
    "columns": Array [
      Array [
        "x",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
      ],
      Array [
        "openshiftRuntimeTime",
        0,
        0,
        0,
        3,
        0,
        0,
      ],
    ],
    "names": Object {
      "openshiftRuntimeTime": "RHOCP Instance Hours",
    },
  },
  "openshiftVcpuTime": Object {
    "colors": Object {
      "openshiftVcpuTime": "#39a5dc",
    },
    "columns": Array [
      Array [
        "x",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
      ],
      Array [
        "openshiftVcpuTime",
        null,
        null,
        null,
        null,
        null,
        null,
      ],
    ],
    "names": Object {
      "openshiftVcpuTime": "RHOCP Core Hours",
    },
  },
  "rhelInstances": Object {
    "colors": Object {
      "rhelInstances": "#ec7a08",
    },
    "columns": Array [
      Array [
        "x",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
      ],
      Array [
        "rhelInstances",
        1,
        1,
        2,
        2,
        2,
        2,
      ],
    ],
    "names": Object {
      "rhelInstances": "RHEL Instances",
    },
  },
  "rhelMemoryTime": Object {
    "colors": Object {
      "rhelMemoryTime": "#ec7a08",
    },
    "columns": Array [
      Array [
        "x",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
      ],
      Array [
        "rhelMemoryTime",
        null,
        null,
        null,
        null,
        null,
        null,
      ],
    ],
    "names": Object {
      "rhelMemoryTime": "RHEL GB Memory Hours",
    },
  },
  "rhelRuntimeTime": Object {
    "colors": Object {
      "rhelRuntimeTime": "#ec7a08",
    },
    "columns": Array [
      Array [
        "x",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
      ],
      Array [
        "rhelRuntimeTime",
        15,
        1,
        3,
        3,
        3,
        3,
      ],
    ],
    "names": Object {
      "rhelRuntimeTime": "RHEL Instance Hours",
    },
  },
  "rhelVcpuTime": Object {
    "colors": Object {
      "rhelVcpuTime": "#ec7a08",
    },
    "columns": Array [
      Array [
        "x",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
        "Invalid date",
      ],
      Array [
        "rhelVcpuTime",
        null,
        null,
        null,
        null,
        null,
        null,
      ],
    ],
    "names": Object {
      "rhelVcpuTime": "RHEL Core Hours",
    },
  },
  "tooltips": Object {
    "format": Object {
      "title": [Function],
    },
    "tooltip": Object {
      "grouped": false,
    },
  },
}
`;

exports[`GraphHelpers should have a specific convertGraphData methods for tooltip titles: tooltip title 1`] = `"Data Lorem Ipsum"`;

exports[`GraphHelpers should have defined methods: calculateGraphTotals keys 1`] = `
Array [
  "instancesOpenshift",
  "instancesRhel",
  "openshiftMemoryTime",
  "openshiftRuntimeTime",
  "openshiftVcpuTime",
  "rhelMemoryTime",
  "rhelRuntimeTime",
  "rhelVcpuTime",
]
`;

exports[`GraphHelpers should have defined methods: convertGraphData keys 1`] = `
Array [
  "openshiftInstances",
  "rhelInstances",
  "openshiftMemoryTime",
  "openshiftRuntimeTime",
  "openshiftVcpuTime",
  "rhelMemoryTime",
  "rhelRuntimeTime",
  "rhelVcpuTime",
  "tooltips",
]
`;
